File: cs125/projects/histo.py
#File: histo.py
#Date: December 7, 2017
#Author: Adam Abad
#Purpose: To construct a histogram for data in a file

from math import *

def intro():

    print()
    print("Program to print a histogram of scores.")
    print("It also produces the high, low, and average score")
    print("along with the standard deviation.")
    print("Written by Adam Abad.")
    print()

def getFile():

    infileName = input("Enter the name of the input file: ")
    print() #forturnin
    print()
    infile = open(infileName, "r")
    return infile

def makeList(file):
    
    mainList = []
    for line in file:
        mainList = mainList + line.split()
    return mainList

def calcScores(scores):

    numScores = []
    high = 0 #Number will always be higher than 0
    low = 100 #Number will always be lower than 100
    total = 0
    for scoreValue in range(1,len(scores),2):
        numScores = numScores + [int(scores[scoreValue])]
    for value in numScores:
        total = total + value
        if value < low:
            low = value
        if value > high:
            high = value
    average = total / len(numScores) #collin's idea

    #deviation calc
    sigma = 0
    for Dvalue in numScores:
        sigma = sigma + ((average - Dvalue) ** 2)
    denom = len(numScores) - 1
    standard = sqrt(sigma / denom)

    return low, high, average, standard, numScores

def output(scores):
    
    print("Name              Score")
    print("====              =====")
    print()
    for index in range(0,len(scores),2):
        print("{0:<12}{1:11.2f}".format(scores[index],float(scores[index + 1])))

def histogram(numScores,low,high):
    
    print("Histogram:")
    print("==========")
    print()
    numScores.sort()
    position = 0
    for histo in range(low,high+1):
        
        print("{0:4}".format(str(histo)+":"), end='')
        while numScores[position] == histo:
            print("*",end='')
            if len(numScores)-1 == position: #No list out of range @ last num
                break
            position = position + 1
            
        print()
    
def main():
    
    intro()
    file = getFile()
    scores = makeList(file)

    file.close()
    low, high, average, standard, numScores = calcScores(scores)
    print("Worst score:{0:13.2f}".format(low))
    print("Best score:{0:14.2f}".format(high))
    print("Average score:{0:11.2f}".format(average))
    print("Standard deviation:{0:6.2f}".format(standard))
    print()

    output(scores)
    print()
    histogram(numScores,low,high)
    print()
main()

File: cs126/projects/Histogram/Histo.java
// File: Histo.java
// Date: Febuary 27, 2018
// Author: Adam Abad
// Purpose: To make and display a histogram of grades

import java.io.*;
import java.util.Scanner;

public class Histo
{
    public static void main(String[] args)
	throws FileNotFoundException
    {	    
        
	intro();
	String[] nameArr;
	double[] scoreArr;
	final int MAX_SCORES = 1000;

	nameArr = new String[MAX_SCORES];
	scoreArr = new double[MAX_SCORES];

	int scores = fillArraysFile(nameArr, scoreArr);

	scoreArr = DoubleArrayOps.subArray(scoreArr,0,scores-1);
	
	double high = getHigh(scoreArr);
	double low = getLow(scoreArr);
	double average = DoubleArrayOps.average(scoreArr);
	double stdDev = DoubleArrayOps.standardDev(scoreArr);

	//Table Output
	summary(high,low,average,stdDev,scoreArr,nameArr);

	//Histogram
	selectionSort(scoreArr);
	printHist(scoreArr);
	System.out.println();
    }
    


    public static void intro()
    {
	System.out.println();
	System.out.println("Program to print a histogram of a "
			   + "list of scores.");
	System.out.println("It also produces the high, low, and "
			   + "average score.");
	System.out.println("along with the standard deviation.");
	System.out.println("Written by Adam Abad.");
	System.out.println();
    }

    public static double getHigh(double[] scoreArr)
    {
	double high = scoreArr[0];
	int numValue = scoreArr.length;
	if(numValue > 0)
	{
	    for(int i = 1; i < numValue; i++)
	    {
		double compare = scoreArr[i];
		if(compare > high)
		    high = compare;
	    }
	}
	else
	{
	    System.out.println("No Numbers Found.");
		return 0.0;
	  
	}
	return high;
    }
    
    public static double getLow(double[] scoreArr)
    {
	double low = scoreArr[0];
	int numValue = scoreArr.length;
	if(numValue > 0)
	{
	    for(int i = 1; i < numValue; i++)
	    {
		double compare = scoreArr[i];
		if(compare < low)
		    low = compare;
	    }
	}
	else
	{
	    System.out.println("No Numbers Found.");
		return 0.0;
	  
	}
	return low;
    }
    
    

    public static void summary(double high, double low,double avg,
			       double stdDev,double[] scoreArr,
			       String[] nameArr)
    {
	System.out.printf("Worst score: %13.2f\n",low);
	System.out.printf("Best score: %14.2f\n",high);
	System.out.printf("Average score: %11.2f\n",avg);
	System.out.printf("Standard deviation: %6.2f\n",stdDev);

	System.out.println();
	System.out.println("Name              Score");
	System.out.println("====              =====");
	System.out.println();

	for(int i = 0; i < scoreArr.length; i++)
	    System.out.printf("%-17s%6.2f\n",nameArr[i],
			       scoreArr[i]);
	System.out.println();
    }

    //Swap and SelectionSort from sheet
    
    public static void swap(double[] a, int i, int j)
    {
	if (i != j)
	{
	    double temp = a[i];
	    a[i] = a[j];
	    a[j] = temp;
	}
    }

    public static void selectionSort(double[] a)
    {
	for(int i = 0; i < a.length - 1; i++)
	{
	    int smallest = i;
	    for(int j = i + 1; j < a.length; j++)
	    {
		if(a[j] < a[smallest])
		{
		    smallest = j;
		}
	    }
	    swap(a, i, smallest);
	}
    }

    public static void printHist(double[] sorted)
    {
	System.out.println("Histogram:");
	System.out.println("==========");
	System.out.println();
	int position = 0;
	for(int x = (int)sorted[0]; x <= (int)sorted[sorted.length - 1];
	    x++)
	{
	    System.out.print(x + ": ");
	    while(x == sorted[position])
	    {
		System.out.print("*");
		position++;
		if(position >= sorted.length - 1)
		    break;
	    }
	    System.out.println();
	}
    }
    
    public static int fillArraysFile(String[] nameArr,
				     double[] scoreArr)
	throws FileNotFoundException
    {
	
	Scanner theKeyboard = new Scanner(System.in);
    
	System.out.print("Enter the name of the scores file: ");
	String filename = theKeyboard.next();
	System.out.println();
	System.out.println(); //Turnin

	Scanner theFile = new Scanner(new File(filename));

	int numscores = 0;
	String name;
	double score;

	while(true)
	{
	    name = theFile.next();
	    if(name.equals("done"))
		break;
	    score = theFile.nextDouble();

	    nameArr[numscores] = name;
	    scoreArr[numscores] = score;

	    numscores++;
	}
	return numscores;
    }
}
